aplt/client.py

```diff
@@ -0,0 +1,206 @@
+        self.processor.handle(dict(messageType="connect", client=self))
+
+    def onMessage(self, payload, isBinary):
+        try:
+            data = json.loads(payload)
+        except Exception as exc:
+            self.processor.handle(dict(messageType="error", exception=exc))
+        else:
+            self.processor.handle(data)
+
+    def onClose(self, wasClean, code, reason):
+        self.harness.remove_client(self)
+
+        try:
+            self.processor.handle(dict(messageType="disconnect",
+                                       was_clean=wasClean, code=code,
+                                       reason=reason))
+        except AttributeError:
+            pass
+
+
+class CommandProcessor(object, policies.TimeoutMixin):
+    """Created per Virtual Client to run a client scenario"""
+    valid_commands = ["connect", "disconnect", "register", "hello",
+                      "unregister", "send_notification", "expect_notification",
+                      "ack", "wait"]
+    valid_handlers = ["connect", "disconnect", "error", "hello",
+                      "notification", "register", "unregister"]
+
+    def __init__(self, scenario, harness):
+        self._harness = harness
+        self._scenario = scenario()
+
+        # Command processing
+        self._last_command = None
+        self._expecting = None
+        self._waiting = False
+
+        # Websocket attributes
+        self._connected = False
+        self._ws_client = None
+        self._notifications = []
+
+    def run(self):
+        """Start the scenario"""
+        self._run_safely(lambda: self._scenario.next())
+
+    def _send_command_result(self, result):
+        self._run_safely(lambda: self._scenario.send(result))
+
+    def _run_safely(self, func, throw=False):
+        try:
+            self._run_command(func())
+        except StopIteration:
+            self._harness.remove_processor(self)
+        except:
+            log.err()
+            self._harness.remove_processor(self)
+
+    def _run_command(self, command):
+        log.msg("Running command: ", command)
+        command_name = command.__class__.__name__
+        try:
+            if command_name not in self.valid_commands:
+                raise Exception("Invalid command: %s" % command_name)
+
+            self._last_command = command_name
+            getattr(self, command_name)(command)
+        except:
+            # Log the exception and shutdown the client
+            log.err()
+            self._harness.remove_processor(self)
+
+    def connect(self, command):
+        """Run the connect command to start the websocket connection"""
+        if self._connected:
+            raise Exception("Already connected")
+        self._harness.connect(self)
+
+    def disconnect(self, command):
+        """Close the websocket connection"""
+        if not self._ws_client:
+            raise Exception("Not connected")
+        self._ws_client.sendClose()
+
+    def hello(self, command):
```

Comment:
for commands like this, do we want to pass additional arguments (e.g. what we do for bridges)? We should properly ignore them, but we should also drop abusive stuff, right?